### ISSUE TITLE
Exception: Cannot find VBoxPython module (tried: VBoxPython3_6, VBoxPython3, VBoxPython)

### ISSUE THREAD

**Issue by printdhruv** [sethmlarson/virtualbox-python#116]
##### ENVIRONMENT

- Operating System: macOS High Sierra 10.13.6
- Python version: Python 3.6.5 |Anaconda, Inc.| (default, Apr 26 2018, 08:42:37) [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin- VirtualBox version: 
- VirtualBox SDK version: /Applications/VirtualBox.app/Contents/MacOS/sdk/- Location where VirtualBox SDK is installed:/Applications/VirtualBox.app/Contents/MacOS/sdk/installer
 
- pyvbox version:1.3.2
- [Not Sure] Happens in latest `master` branch?

##### SUMMARY

It doesn't work even after setting up paths.
##### STEPS TO REPRODUCE

```
import virtualbox
vbox = virtualbox.VirtualBox()
```
##### ACTUAL RESULTS
```
m=VBoxPython3_6 x=No module named 'VBoxPython3_6'
m=VBoxPython3 x=No module named 'VBoxPython3'
m=VBoxPython x=No module named 'VBoxPython'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/anaconda3/lib/python3.6/site-packages/virtualbox/library_ext/vbox.py", line 22, in __init__
    manager = virtualbox.Manager()
  File "/anaconda3/lib/python3.6/site-packages/virtualbox/__init__.py", line 143, in __init__
    self.manager = vboxapi.VirtualBoxManager(mtype, mparams)
  File "/anaconda3/lib/python3.6/site-packages/vboxapi/__init__.py", line 989, in __init__
    self.platform = PlatformXPCOM(dPlatformParams)
  File "/anaconda3/lib/python3.6/site-packages/vboxapi/__init__.py", line 750, in __init__
    import xpcom.vboxxpcom
  File "/Applications/VirtualBox.app/Contents/MacOS/sdk/bindings/xpcom/python/xpcom/vboxxpcom.py", line 78, in <module>
    raise Exception('Cannot find VBoxPython module (tried: %s)' % (', '.join(_asVBoxPythons),))
Exception: Cannot find VBoxPython module (tried: VBoxPython3_6, VBoxPython3, VBoxPython)
```

**Comment by SethMichaelLarson** [comment]
This might be a good place to start as this issue seems related: https://github.com/SethMichaelLarson/virtualbox-python/issues/102

**Comment by printdhruv** [comment]
So, what's the resolution there? You are working on that !
I tired that dirty hack but still doesn't help.
Would you like to suggest some other hack to solve this issue?

**Comment by SethMichaelLarson** [comment]
Could you provide the output of: `nm -D [Wherever VirtualBox.so files are installed] | grep PyInit` to help me understand why the dynamic library isn't loading? I don't own a macOS system to try this on, unfortunately.

**Comment by uohoo** [comment]
Hi, I have the same problem using Ubuntu18.04 and Python3.6
When I try:
```
import virtualbox
vbox = virtualbox.VirtualBox()
```
The error:
```
Cannot find VBoxPython module (tried: VBoxPython3_6m, VBoxPython3m, VBoxPython)
  File "/home/u234/&lt;frozen importlib._bootstrap&gt;", line 219, in _call_with_frames_removed
  File "/home/u234/&lt;frozen importlib._bootstrap_external&gt;", line 678, in exec_module
  File "/home/u234/&lt;frozen importlib._bootstrap&gt;", line 665, in _load_unlocked
  File "/home/u234/&lt;frozen importlib._bootstrap&gt;", line 955, in _find_and_load_unlocked
  File "/home/u234/&lt;frozen importlib._bootstrap&gt;", line 971, in _find_and_load
  File "/home/u234/vbox_test.py", line 2, in <module>
    vbox = virtualbox.VirtualBox()
```

My installation folder is: /usr/lib/virtualbox, where I can find the file VBoxPython.so. 
Where are these searched?

Thank you.

**Comment by Kentzo** [comment]
@SethMichaelLarson I think it's not loading because VirtualBox.app on macOS is shipped only with `VBoxPython2_6.so` and `VBoxPython2_7.so`.

**Comment by printdhruv** [comment]
Can you guy develop something automatic way? This thing is pain. Why don't you can go for something which finds / install things itself vs we have to do it?

**Comment by SethMichaelLarson** [comment]
@printdhruv [Above I asked for information](https://github.com/SethMichaelLarson/virtualbox-python/issues/116#issuecomment-421443371) from you to better understand your case but you didn't reply. Could you response to my question so I can attempt to fix your case as well?

**Comment by Kentzo** [comment]
@printdhruv I suppose it should be possible to develop a plugin that uses C-API over XPCOM instead of VBoxPython. That it will be a pain to maintain as well.

If you have control over the deployment (what version of VirtualBox with what version of Python), you can compile VirtualBox yourself with bindings for you Python.

**Comment by nixjdm** [comment]
I'm experiencing the same thing, and I'm all but certain this is a duplicate of #102, just with python 3.6 instead of 3.5.

```bash
cd /usr/lib/virtualbox/
sudo cp VBoxPython3_6m.so VBoxPython3_6.so
```

'fixes' the problem.

How can I poke the virtualbox devs about this? https://www.virtualbox.org/ticket/17448 looks to have no activity.

**Comment by SethMichaelLarson** [comment]
@nixjdm I'm wondering the same thing. I haven't heard anything on the issue or otherwise.

**Comment by franz101** [comment]
Not for Mac though:
VBoxPython3_7m.so: unknown file type, first eight bytes: 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x03

**Comment by sethmlarson** [comment]
Hm, maybe that's something to raise on the VirtualBox forums?

**Comment by nkcr** [comment]
> You need to somehow compile VBoxPython for your Python first before you can use the bindings.

Any idea how to do this?

@Kentzo  is right, the MacOS package comes only with the `VBoxPython2_6.so` and `VBoxPython2_7.so` versions...

**Comment by ciarancourtney** [comment]
@sethmlarson how are you installing VirtualBox? As far as I can see all the latest ppa/deb install methods only install with Python 2.7 bindings incl buster (which I thought dropped python2?)

So ATM a official PPA install of VirtualBox 6.0.12 on Ubuntu 18.04 will not work with Python 3+

**Comment by ciarancourtney** [comment]
I tried dropping `VBoxPython3_6m.so` from the EL8 rpm beside `VBoxPython2_7.so` but I still get the same error `Cannot find VBoxPython module (tried: VBoxPython3_6m, VBoxPython3m, VBoxPython)`, even when removing the `m`

FYI

    $ nm -D VBoxPython3_6m.so | grep PyInit
    000000000001268f T PyInit_VBoxPython3_6m

**Comment by ciarancourtney** [comment]
FYI As a workaround I created a [Dockerfile](https://github.com/ciarancourtney/build-virtualbox-docker) to build a minimal VirtualBox from source on bionic, then you can copy the two required files to your local install.

    sudo docker cp vbox-build:/VirtualBox/out/linux.amd64/release/obj/VBoxPython/VBoxPython.so /usr/lib/virtualbox/
    sudo docker cp vbox-build:/VirtualBox/out/linux.amd64/release/obj/VBoxPython3_6m/VBoxPython3_6m.so /usr/lib/virtualbox/

**Comment by Ryuchen** [comment]
> Not for Mac though:
> VBoxPython3_7m.so: unknown file type, first eight bytes: 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x03

did you solve this problem?

**Comment by HengyueLi** [comment]
what about you, do you solve this? @Ryuchen

**Comment by Ryuchen** [comment]
> what about you, do you solve this? @Ryuchen

No, It seems we have to compiler  it by ourself

**Comment by ekorian** [comment]
I'm getting the same error

**Comment by hgoertz82** [comment]
Has anyone been able to get it to run on MacOs?
I still get the error described here.

**Comment by AlexHartford** [comment]
Facing the same issue on Ubuntu 20.04. Tried 

> ```shell
> cd /usr/lib/virtualbox/
> sudo cp VBoxPython3_6m.so VBoxPython3_6.so
> ```

In my case it was VBoxPython3_7m, but still, that was unable to solve the issue for me. Can anyone please help?

**Comment by vidhanio** [comment]
Has there been any fix to this issue on macos? No matter what, virtualbox only has 2_7 `.so`s in `Virtualbox.app/Contents/MacOS/`, and cannot use virtualbox-python in python3 :(

**Comment by sethmlarson** [comment]
The fix isn't in this library, needs to be implemented upstream. See the link to the virtualbox forums.

**Comment by vidhanio** [comment]
Which link are you talking about? Is it [this one](https://www.virtualbox.org/ticket/17448)?

**Comment by vidhanio** [comment]
darn. so looks like i wont be able to use virtualbox-python until they decide to implement py3? :/ looks like ill have to look into running a vm to dev with it then, thanks for the help!

**Comment by sethmlarson** [comment]
@vidhanio It's not a Python 3 issue, it's about the binaries Virtualbox ships not being aware of ABI flags. If your Python version doesn't have ABI flags it'll work fine.

**Comment by vidhanio** [comment]
This seems to be way over my head, is it possible to have a step-by-step instruction to this?

**Comment by alainpannetier** [comment]
For those who still have the problem., I manage to solve it by scavenging the **VBoxPython3_8.so** from the Ubuntu 6.26 package.
It is not in the Oracle PPA package, but somehow, you can find it in the slower ubuntu package.

Then: 
```
sudo cp VBoxPython3_8.so /usr/lib/virtualbox/
```

**Comment by craigatfetch** [comment]
I have this issue on MacOS Monterey:

`Exception: Cannot find VBoxPython module (tried: VBoxPython3_9, VBoxPython3, VBoxPython)`

I installed VirtualBox from the Oracle site, not with homebrew.
Any hints?

**Comment by nvictor** [comment]
Wondering 🙂

Does anyone have any insight into building the VirtualBox COM Python module on the mac? Using Python 3 this time? This seems to be the recurring issue mentioned over and over here.

I have had this line from the source code bookmarked for many years now: https://www.virtualbox.org/browser/vbox/trunk/src/VBox/Installer/darwin/Makefile.kmk#L481

The maintainers do not seem interested in making the changes we all want here. So maybe with enough knowledge we can start a community supported module.

**Comment by SergiMartinezR** [comment]
Hello, I'm having the same issue. I edited the vboxxpcom.py file to see what the error was (I have the modules mentioned in the exception), and I figured out that it is searching the module in the same directory I'm working on.
So I copied the VBoxPython3.so module to the same dir as my script and the error was gone.

I guess I did something wrong (maybe my path is incomplete), cause I don't think this is the way it has to be done.

If it's useful I'm working on Arch in a venv.
